File: spring-web/src/main/java/com/example/springwebex/model/prop/MyDataSourceProperties.java
package com.example.springwebex.model.prop;

import java.time.Duration;
import java.util.List;

import org.hibernate.validator.constraints.time.DurationMax;
import org.hibernate.validator.constraints.time.DurationMin;
import org.springframework.boot.context.properties.ConfigurationProperties;
import org.springframework.validation.annotation.Validated;

import jakarta.validation.constraints.Max;
import jakarta.validation.constraints.Min;
import jakarta.validation.constraints.NotEmpty;
import lombok.Data;

@Data
@Validated
@ConfigurationProperties("my.datasource")
public class MyDataSourceProperties {

    @NotEmpty
    private String url;

    @NotEmpty
    private String username;

    @NotEmpty
    private String password;

    private Etc etc;

    @Data
    public static class Etc {
        @Min(1)
        @Max(999)
        private int maxConnection;

        @DurationMin(seconds = 1)
        @DurationMax(seconds = 60)
        private Duration timeout;

        private List<String> options;
    }

}

File: spring-web-kotlin/src/main/kotlin/com/example/springwebex/model/prop/MyDataSourceProperties.kt
package com.example.springwebex.model.prop

import java.time.Duration
import org.hibernate.validator.constraints.time.DurationMax
import org.hibernate.validator.constraints.time.DurationMin
import org.springframework.boot.context.properties.ConfigurationProperties
import org.springframework.validation.annotation.Validated
import jakarta.validation.constraints.Max
import jakarta.validation.constraints.Min
import jakarta.validation.constraints.NotEmpty

@Validated
@ConfigurationProperties("my.datasource")
data class MyDataSourceProperties(
    @NotEmpty
    var url: String = "",

    @NotEmpty
    var username: String = "",

    @NotEmpty
    var password: String = "",

    var etc: Etc? = null
) {
    data class Etc(
        @Min(1)
        @Max(999)
        var maxConnection: Int = 1,

        @DurationMin(seconds = 1)
        @DurationMax(seconds = 60)
        var timeout: Duration? = null,

        var options: List<String>? = null
    )
}
